[PatternMatching] Add generic API for matching constants using custom conditions

goldsteinn opened the pull request:
- **[PatternMatching] Add generic API for matching constants using custom conditions**
- **[InstCombine] Add example usage for new `Checked` matcher API**

The new API is:
    `m_CheckedInt(Lambda)`/`m_CheckedFp(Lambda)`
        - Matches non-undef constants s.t `Lambda(ele)` is true for all
          elements.
    `m_CheckedIntAllowUndef(Lambda)`/`m_CheckedFpAllowUndef(Lambda)`
        - Matches constants/undef s.t `Lambda(ele)` is true for all
          elements.

The goal with these is to be able to replace the common usage of:
```
    match(X, m_APInt(C)) && CustomCheck(C)
```
with
```
    match(X, m_CheckedInt(C, CustomChecks);
```

The rationale if we often ignore non-splat vectors because there are
no good APIs to handle them with and its not worth increasing code
complexity for such cases.

The hope is the API creates a common method handling
scalars/splat-vecs/non-splat-vecs to essentially make this a
non-issue.

goldsteinn committed:
[PatternMatching] Add generic API for matching constants using custom conditions

The new API is:
    `m_CheckedInt(Lambda)`/`m_CheckedFp(Lambda)`
        - Matches non-undef constants s.t `Lambda(ele)` is true for all
          elements.
    `m_CheckedIntAllowUndef(Lambda)`/`m_CheckedFpAllowUndef(Lambda)`
        - Matches constants/undef s.t `Lambda(ele)` is true for all
          elements.

The goal with these is to be able to replace the common usage of:
```
    match(X, m_APInt(C)) && CustomCheck(C)
```
with
```
    match(X, m_CheckedInt(C, CustomChecks);
```

The rationale if we often ignore non-splat vectors because there are
no good APIs to handle them with and its not worth increasing code
complexity for such cases.

The hope is the API creates a common method handling
scalars/splat-vecs/non-splat-vecs to essentially make this a
non-issue.

dtcxzyw commented on llvm/include/llvm/IR/PatternMatch.h:
```diff
@@ -384,16 +387,17 @@ struct cstval_pred_ty : public Predicate {
 /// specialization of cstval_pred_ty for ConstantInt
-template <typename Predicate>
-using cst_pred_ty = cstval_pred_ty<Predicate, ConstantInt>;
+template <typename Predicate, bool AllowUndefs = true>
```
I would make `AllowUndefs` explicit or default to false.

goldsteinn commented:
ping

dtcxzyw commented:
Please check the compile-time impact of this change.

goldsteinn commented:
> Please check the compile-time impact of this change.

https://llvm-compile-time-tracker.com/compare.php?from=817f453aa576286aaca0a6b0244e6ab08516b80c&to=05871ee629407932bc6576224dcc2ae99db473fa&stat=instructions:u

Think basically no impact, maybe a slight regression stage2-O0 and slight improvement stage2-O3, but looks to be within err range and a bit scattered.

goldsteinn commented:
ping

dtcxzyw commented on llvm/include/llvm/IR/PatternMatch.h:
```diff
@@ -452,30 +470,93 @@ template <typename Predicate> struct apf_pred_ty : public Predicate {
+inline api_or_undef_pred_ty<custom_checkfn<APInt>>
+m_CheckedIntAllowUndef(const APInt *&V,
+                       function_ref<bool(const APInt &)> CheckFn) {
+  api_or_undef_pred_ty<custom_checkfn<APInt>> P(V);
```
Can we pass `CheckFn` into the constructor?

goldsteinn replied:
No (other functions also do it like this).

arsenm replied:
They might, but why?

goldsteinn replied:
Maybe there is a syntax for it, but don't know the syntax for constructing fields in inherited + base type.

dtcxzyw commented on llvm/include/llvm/IR/PatternMatch.h:
```diff
@@ -452,30 +470,93 @@ template <typename Predicate> struct apf_pred_ty : public Predicate {
 //
 ///////////////////////////////////////////////////////////////////////////////
 
+template <typename APTy> struct custom_checkfn {
```
It would be better to add some unit tests for new helpers.

goldsteinn commented:
ping

goldsteinn commented:
> I'd like to defer this until after #88217, which would address my primary concern with this (the ever present undef footgun).

Actually any issue getting just the default APIs (not supporting `undef`) in now? Don't imagine it will get in your way. If you think it will be a problem though, this can ofc be deferred.

goldsteinn committed:
[InstCombine] Add non-splat test for `(icmp (lshr x, y), x)`; NFC

goldsteinn committed:
[InstCombine] Add example usage for new Checked matcher API

There is no real motivation for this change other than to highlight a
case where the new `Checked` matcher API can handle non-splat-vecs
without increasing code complexity.

goldsteinn commented:
rebased

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:
```diff
@@ -7144,33 +7144,37 @@ Instruction *InstCombinerImpl::foldICmpCommutative(ICmpInst::Predicate Pred,
+    if (match(Op0,
+              m_UDiv(m_Specific(Op1), m_CheckedInt([](const APInt &Divisor) {
+                       return Divisor.ugt(1);
+                     })))) {
```
This is better than some alternatives, but this still makes the code significantly uglier exclusively for the sake of non-splat vectors. I'm not entirely convinced that this is a good tradeoff.

goldsteinn replied:
I don't think it would be ugly if we made the lambda a helper. Ill do so.

nikic commented on llvm/include/llvm/IR/PatternMatch.h:
```diff
@@ -460,6 +460,39 @@ template <typename Predicate> struct apf_pred_ty : public Predicate {
+  bool isValue(const APTy &C) { return CheckFn(C); }
+};
+
+// Match and integer or vector where CheckFn(ele) for each element is true.
+// For vectors, poison elements are assumed to match.
```
```suggestion
/// Match and integer or vector where CheckFn(ele) for each element is true.
/// For vectors, poison elements are assumed to match.
```

nikic commented on llvm/include/llvm/IR/PatternMatch.h:
```diff
@@ -460,6 +460,39 @@ template <typename Predicate> struct apf_pred_ty : public Predicate {
+  return P;
+}
+
+// Match and float or vector where CheckFn(ele) for each element is true.
+// For vectors, poison elements are assumed to match.
```
```suggestion
/// Match and float or vector where CheckFn(ele) for each element is true.
/// For vectors, poison elements are assumed to match.
```

nikic commented on llvm/test/Transforms/InstCombine/icmp-div-constant.ll:
```diff
@@ -421,6 +421,16 @@ define <4 x i1> @lshr_by_const_cmp_sle_value(<4 x i32> %x) {
+  %v = lshr <4 x i32> %x, <i32 3, i32 3, i32 3, i32 5>
+  %r = icmp sle <4 x i32> %v, %x
+  ret <4 x i1> %r
+}
```
Also add tests for the other three cases you modified?

nikic commented on llvm/unittests/IR/PatternMatch.cpp:
```diff
@@ -611,6 +611,212 @@ TEST_F(PatternMatchTest, BitCast) {
+  EXPECT_FALSE(CustomCheckZeroF.isValue(F0));
+  EXPECT_TRUE(CustomCheckEqF1.isValue(F0));
+  EXPECT_FALSE(CustomCheckNeF1.isValue(F0));
+}
```
Hm, I'd consider custom_checkfn to be a PatternMatch implementation detail -- do we really need unit tests for it if you already have tests for m_CheckedInt below?

goldsteinn replied:
@dtcxzyw  requested them. I'm ambivalent.

dtcxzyw replied:
> do we really need unit tests for it if you already have tests for m_CheckedInt below?

The version I reviewed https://github.com/llvm/llvm-project/commit/4d7774dfdae01dbc8ebb9b30748cfb8638240ca3 doesn't include tests for m_CheckedInt.

It is ok to remove tests for CustomCheckFn.

goldsteinn replied:
Okay, dropped these tests.

goldsteinn committed:
done

nikic reviewed (approved):
LG

goldsteinn commented on llvm/test/Transforms/InstCombine/icmp-div-constant.ll:
```diff
@@ -421,6 +442,28 @@ define <4 x i1> @lshr_by_const_cmp_sle_value(<4 x i32> %x) {
+  %v = ashr <4 x i32> %x, <i32 1, i32 2, i32 3, i32 4>
+  %r = icmp sge <4 x i32> %v, %x
+  ret <4 x i1> %r
+}
```
Already had non-splat negative tests (i.e `udiv_x_by_const_cmp_eq_value_neg`). In retrospect probably should have had those tests be scalars as they didn't really check the old logic, but we can use them here.

nikic commented on llvm/include/llvm/IR/PatternMatch.h:
```diff
@@ -460,6 +460,39 @@ template <typename Predicate> struct apf_pred_ty : public Predicate {
+}
+
+inline api_pred_ty<custom_checkfn<APInt>>
+m_CheckedInt(const APInt *&V, function_ref<bool(const APInt &)> CheckFn) {
```
Looking at this again just now, what's the point of these methods returning a `const APInt *`? In that case, isn't this limited to splats anyway and you could just as well use `m_APInt` + a check and save you the lambda?

I think this should be providing the result as a `Constant *`, which would support non-splat values matching the predicate...

goldsteinn replied:
Missed this... yeah think you are right. Should be a constant.

goldsteinn replied:
Interesting the norm case seems to be to always export these as `APInt`, if you looks at something like `m_Negative`/`m_NonNegative` its the same deal.

Any objection to me replacing all of these?

nikic replied:
Assuming there are not a lot of uses of them, replacing them should be fine. Something to be careful about is that the direct replacement here would be m_APIntAllowPoison, not m_APInt.

If there are many uses, it may make sense to retain them as a convenience (I don't think m_CheckedInt with APInt makes sense as a convenience though, it's strictly more complicated than just writing out the check).

goldsteinn replied:
See: https://github.com/llvm/llvm-project/pull/91377 decided not to do the other. Didn't seem like much upside and made code less compact.
